proposals/fibers/Explainer.md

```diff
@@ -0,0 +1,874 @@
+
+`handle` blocks may be nested; in which case, the available `event` blocks are the concatenation of all the `event` blocks in all the `handle` blocks that a given switching instruction is textually enclosed within&mdash;up to and including the function body itself.
+
+### Fiber instructions
+
+#### `fiber.new` Create a new fiber
+
+The `fiber.new` instruction creates a new fiber. The instruction has a literal operand which is the index of a function of type `[[ref fiber  r*] t*]->[r*]`, together with corresponding values on the argument stack.
+
+The result of the `fiber.new` instruction is a `fiber` which is the identifier for the newly created fiber. The identity of the fiber is also the first argument to the fiber function itself&mdash;this allows fibers to know their own identity in a straightforward way.
+
+The return values that the fiber function returns represent the return values of the fiber. This is reflected in the type signatures of the `fiber` and the function.
+
+The new fiber is created in a suspended state: it must be the case that the function referenced has one or more `event` blocks within the top-level block of that function. When the fiber is resumed, one of those `event` blocks will be entered as a result of the `fiber.resume`&mdash;depending on the actual event used to resume the fiber.
+
+>If a function is referenced in a `fiber.new` instruction has no `event` blocks at the toplevel, then this function will cause a trap when attempting to resume the new fiber&mdash;because, without any `event` blocks, the fiber will not be able to respond to the resume.
+
+#### `fiber.spawn` Create and enter a new fiber
+
+The `fiber.spawn` instruction creates a new fiber and immediately enters it. It requires a literal function operand of type `[[fiber r*] t*] -> [r*]`, together with values corresponding to `t*` on the value stack.
+
+The `fiber.spawn` instruction is analogous to a `call` instruction, the effect of the instruction is to leave on the value stack the return values of the fiber's function. However, as with other fibers, the new fiber may suspend; in which case there should be appropriate `event` blocks in the instruction stream enclosing the `fiber.spawn` instruction.
+
+Notice that only the newly created fiber has access to the reference that identifies the new fiber. If it is desired that the parent has access to this the code generator would typically insert a `fiber.suspend` instruction at the start of the function and pass out the `fiber` in an appropriate event description.
+
+#### `fiber.suspend` Suspend an active fiber
+
+The `fiber.suspend` instruction takes a fiber as an argument and suspends the fiber. The identified fiber must either be the `active` fiber, or a resume ancestor of the active fiber. 
```

Comment:
You absolutely do want to be able to suspend a resume ancestor. This shows up, for example, when combining different use cases. E.g., an async function/generator combo. You absolutely don't want to have to search the stack looking for the right place to suspend/resume - dynamic scoping brings unfortunate collisions.

Reply:
Wrt stack state: for a number of engine-safety related reasons, whenever stacks are switched, all the necessary register state is stored during that switch. The total cost is approximately equivalent to a function call (+ mutex in some cases).

Reply:
It seems better to walk the stack (of fibers) when suspending so that stack pointers can be stashed than to have to eagerly stash and restore stack pointers even when there might not be any suspension.

Reply:
You need to, if for no other reason than to ensure safety of GC.
However, this is not in any case the major cost for switching stacks. That mutex is.